### PR TITLE
Create custom Tooltip that wraps MUI component and use on playback controls

### PR DESCRIPTION
**Pull request by 2metres**
**User-Facing Changes**
Playback controls now include keyboard shortcut

**Description**
Alternative approach to https://github.com/foxglove/studio/pull/7307

### CODE REVIEW

File: packages/studio-base/src/components/Tooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    () => React.cloneElement(children, { onClick: handleChildClick }),
+    [children, handleChildClick],
+  );
+
+  return (
+    <MuiTooltip {...rest} PopperProps={{ ...PopperProps, popperRef }}>
+      {clonedChildren}
+    </MuiTooltip>
+  );
+}
```
Comment (2metres) [inline packages/studio-base/src/components/Tooltip.tsx:38]: We wrap the children elements in a Tooltip and interfere with the popper ref to close it when we click.

**Comment by defunctzombie** [comment]
I am still 👎 on the tooltip showing up so quickly. I find it extremely annoying to use the playback controls and have it instantly pop up. I can't actually think of any time when I want to see the tooltip on playback controls but certainly not immediately.

File: packages/studio-base/src/components/PlaybackControls/index.tsx [hunk 2]
```diff
@@ -230,43 +230,71 @@ export default function PlaybackControls(props: {
...
+                  Seek backward <kbd>←</kbd>
+                </>
+              }
+            >
+              <HoverableIconButton
+                disabled={disableControls}
+                size="small"
+                icon={<Previous20Regular />}
+                activeIcon={<Previous20Filled />}
+                onClick={() => {
```
Comment (defunctzombie) [inline packages/studio-base/src/components/PlaybackControls/index.tsx:247]: We can get rid of this wrapping function:
```suggestion
                onClick={seekBackwardAction
```

File: packages/studio-base/src/components/PlaybackControls/index.tsx [hunk 3]
```diff
@@ -230,43 +230,71 @@ export default function PlaybackControls(props: {
...
+                  Seek forward <kbd>→</kbd>
+                </>
+              }
+            >
+              <HoverableIconButton
+                disabled={disableControls}
+                size="small"
+                icon={<Next20Regular />}
+                activeIcon={<Next20Filled />}
+                onClick={() => {
```
Comment (defunctzombie) [inline packages/studio-base/src/components/PlaybackControls/index.tsx:283]:
```suggestion
                onClick={seekForwardAction
```

**Comment by defunctzombie** [comment]
Tooltip does not seem to be closing on click.

**Comment by 2metres** [comment]
> Tooltip does not seem to be closing on click.

Steps to reproduce?

File: packages/studio-base/src/components/PlaybackControls/index.tsx (hunk 3 above)
Comment (2metres) [reply packages/studio-base/src/components/PlaybackControls/index.tsx:283]: Yeah i think so. Just need to make the callback more flexible with it's types

File: packages/studio-base/src/components/HoverableIconButton.tsx [hunk 4]
```diff
@@ -46,16 +54,25 @@ const HoverableIconButton = forwardRef<HTMLButtonElement, Props>((props, ref) =>
     }
   }, [props.disabled]);
 
+  const iconComponent = useMemo(() => {
+    if (activeIcon != undefined) {
```
Comment (defunctzombie) [inline packages/studio-base/src/components/HoverableIconButton.tsx:58]: If you default `activeIcon` to `icon` when it is not available you can collapse this function to a single line.

From there I am not sure why you need the memo? The icon component would be stable from the caller.

File: packages/studio-base/src/components/PlaybackControls/index.tsx [hunk 5]
```diff
@@ -247,21 +259,13 @@ export default function PlaybackControls(props: {
...
-              enterDelay={1200}
-              enterNextDelay={800}
-              title={
-                <>
-                  {isPlaying ? "Pause" : "Play"} <kbd>Space</kbd>
-                </>
-              }
-            >
+            <Tooltip enterDelay={1200} enterNextDelay={800} title={playPauseButton.title}>
               <HoverableIconButton
```
Comment (defunctzombie) [inline packages/studio-base/src/components/PlaybackControls/index.tsx:263]: I wonder if it would be easier to understand this flow if you have `isPlaying ? <HoverIconButton> : <HoverIconButon>` instead of trying to change the props within one component? Then you can see what the playing state looks like and the paused state quite easily.

File: packages/studio-base/src/components/Tooltip.tsx [hunk 6]
```diff
@@ -3,23 +3,28 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
 import { Tooltip as MuiTooltip, TooltipProps } from "@mui/material";
-import { Instance } from "@popperjs/core";
-import { MouseEvent, useCallback, useMemo, useRef } from "react";
+import { MouseEvent, useCallback, useMemo, useState } from "react";
 
 /**
  * This component exists as result of a team preference to close the
```
Comment (defunctzombie) [inline packages/studio-base/src/components/Tooltip.tsx:9]:
```suggestion
 * Tooltip wraps a MUI tooltip to close the tooltip when the child component is clicked. MUI tooltip does not have this feature.
```

File: packages/studio-base/src/components/Tooltip.tsx [hunk 7]
```diff
@@ -31,7 +36,13 @@ export function Tooltip(props: TooltipProps): JSX.Element {
   );
 
   return (
-    <MuiTooltip {...rest} PopperProps={{ ...PopperProps, popperRef }}>
+    <MuiTooltip
+      {...rest}
+      open={open}
+      onMouseEnter={handleMouseEnter}
```
Comment (defunctzombie) [inline packages/studio-base/src/components/Tooltip.tsx]: If we can have these mouse events on the MuiTooltip can we not also have `onClick`?

**Comment by defunctzombie** [comment]
If I click the button before the tooltip shows up the tooltip shows up after I've clicked the button. :(

File: packages/studio-base/src/components/ShareJsonModal.tsx [hunk 8]
```diff
@@ -108,15 +109,16 @@ export function ShareJsonModal({
...
-          <HoverableIconButton
-            activeColor="error"
-            onClick={() => {
-              setValue("{}");
-            }}
-            title="Clear"
-            aria-label="Clear"
-            icon={<Delete20Regular />}
-          />
+          <Tooltip title="Clear">
```
Comment (defunctzombie) [inline packages/studio-base/src/components/ShareJsonModal.tsx:112]: Please avoid scope creep on the PR - it makes it take longer to review the PR since now we need to also consider these changes. Right now the PR was about changing playback control tooltips.

File: packages/theme/src/components/MuiTooltip.ts [hunk 9]
```diff
@@ -2,25 +2,39 @@
...
+import { Fade } from "@mui/material";
+import tc from "tinycolor2";
 
 import { OverrideComponentReturn } from "../types";
 
 export const MuiTooltip: OverrideComponentReturn<"MuiTooltip"> = {
   defaultProps: {
     arrow: true,
     TransitionComponent: Fade,
+    enterDelay: 800,
```
Comment (defunctzombie) [inline packages/theme/src/components/MuiTooltip.ts:14]: Please add comments here on our reasoning for adjusting these - specifically the enterNextDelay

File: packages/studio-base/src/components/PlaybackControls/index.tsx (hunk 5 above)
Comment (2metres) [reply packages/studio-base/src/components/PlaybackControls/index.tsx:259]: The memo was largely trying to debug some of the other issues and isolate rerenders.

Comment (2metres) [reply packages/studio-base/src/components/PlaybackControls/index.tsx:259]: Reverting to previous version

**Comment by defunctzombie** [comment]
@2metres there are conflicts on the PR that need resolving

**Comment by defunctzombie** [comment]
This is still broken. When I click the play button the tooltip shows up.

When I push the key for the tooltip I would also expect it to go away.
